### ISSUE TITLE
k_nearest strange behavior

### ISSUE THREAD

**Issue by igoronzy** [biocore-ntnu/pyranges#189]
I assumed that k_nearest with k=1 should give the same results as nearest. However, when I apply these functions to a table with 256,596 rows:
nearest: 256,596 
k_nearest with k=1: 498,410
k_nearest with k=1, ties='first': 449,778
k_nearest with k=1, ties='different': 451,379

Why is k_nearest returning more than one row for each row row in my original table?

**Comment by endrebak** [comment]
I will look into it :)

**Comment by endrebak** [comment]
Do you have the possibility of sharing the file?

**Comment by endrebak** [comment]
I've only checked this briefly during lunch (will look at it after work), but it seems like k_nearest keeps both ties in case it overlaps on both strands.

```
In [67]: ndf.head(628).tail(5)
Out[67]:
    Chromosome    Start      End Strand  Start_b    End_b Strand_b  Distance
623       chr1  3840998  3841098      +  3839597  3839697        -      1302
624       chr1  3850891  3850991      +  3851147  3851247        -       157
625       chr1  3853417  3853517      +  3853383  3853483        -         0
626       chr1  3858601  3858701      +  3858265  3858365        +       237
627       chr1  3860347  3860447      +  3860337  3860437        -         0

In [68]: kdf.head(628).tail(5)
Out[68]:
    Chromosome    Start      End Strand  Start_b    End_b Strand_b  Distance
623       chr1  3839742  3839842      +  3839597  3839697        -        46
624       chr1  3840998  3841098      +  3839597  3839697        -      1302
625       chr1  3850891  3850991      +  3851147  3851247        -       157
626       chr1  3853417  3853517      +  3853499  3853599        +         0
627       chr1  3853417  3853517      +  3853383  3853483        -         0
```

**Comment by endrebak** [comment]
I should really add tests for `k_nearest `to see that it preserves some obvious invariants.

Anyways, lunch is over. Will get back to this tonight :)

**Comment by igoronzy** [comment]
Yes, there appears to be an edge case with strandedness. If one of the pyranges is stranded, but the other is not, you still get (+) and (-) results. 
Minimal case:
f1 = pr.from_dict({'Chromosome': ['chr1', 'chr1'], 'Start': [3, 4],'End': [6, 9], 'Strand': ['+', '-']})
f2 = pr.from_dict({'Chromosome': ['chr1'], 'Start': [7], 'End': [9]})

f2.k_nearest(f1) --> 2 results
f2.k_nearest(f1, k=1) --> 2 results
f2.k_nearest(f1, k=1, strandedness=False) --> 2 results (?????)

I was expected the results to ignore strand if only one input was stranded. Should the "strandedness" flag not enforce this?

**Comment by endrebak** [comment]
Thanks! Beautiful work. My daughter was sick today so I've had no energy or time to fix it, but I will do so soon.

I also see that these functions might have been unfinished. If you do `help(f1.k_nearest)` you see that the docs aren't completely finished. Looking at the `git log` it seems like I wrote this code during the last Genome Informatics conference in NY. So I guess I must have forgotten that the code was only 98% done when I got home and pushed it anyways 😳 

Please tell me if you notice other weirdness :)

**Comment by endrebak** [comment]
I'd just like to say that I've started creating tests and fixing these errors :)

**Comment by endrebak** [comment]
https://github.com/biocore-ntnu/pyranges/pull/191

first fix (for when ties=None). I'll look at the other cases soon.
